Ensure env vars don't exist prior to win-env test

dra27 opened the pull request:
The `test_env.ml` test in `tests/lib-unix/win-env` sets environment variables called `FOO` and `FOO2`. I just hit a false-negative testing 4.06.1 because my environment happened to include the environment variable `FOO` from something else I'd been doing.

This tiny tweak ensures that the variables are unset before the test is run - less of an issue as the 4.07 version of this test is slightly different, though still capable of producing a false positive.

nojb commented:
LGTM!

xavierleroy commented:
I have no idea why this PR was never merged.  @dra27, @nojb : if you think this PR is still relevant, could you please merge?

nojb commented:
This PR was against the old `Makefile`-based testsuite, so it cannot be readily merged. Also, my understanding is that `ocamltest` runs tests with an explicitly built environment so this should no longer be an issue.

dra27 commented:
The issues does still exist - as far as I can tell, `ocamltest` has the facility to control the environment but doesn't use it?

@shindere - this issue was originally about an environment variable set before the tests were run affecting the test result. In 4.06, the test expected `FOO` not to be set, and would fail if the test was run with it set; in 4.07, the test was changed - the only thing which could happen is that any one of the environment variable setting functions being tested could fail to work but the test would appear to pass if either `FOO` or `FOO2` had been set beforehand. Is ocamltest supposed to scrub variables out of the environment before running tests?

shindere commented:
David Allsopp (2019/10/15 01:40 -0700):
> @shindere - this issue was originally about an environment variable
> set before the tests were run affecting the test result. In 4.06, the
> test expected `FOO` not to be set, and would fail if the test was run
> with it set; in 4.07, the test was changed - the only thing which
> could happen is that any one of the environment variable setting
> functions being tested could fail to work but the test would appear to
> pass if either `FOO` or `FOO2` had been set beforehand. Is ocamltest
> supposed to scrub variables out of the environment before running
> tests?

No, I don't think ocamltest does something like that.
Shall we add an "unset" directive to its DSL so that it can be
guaranteed in a test that a given variable is not defined?

mshinwell commented:
I would have thought that ocamltest should always run tests in a clean environment (or something very close to that).

shindere commented:
Mark Shinwell (2019/10/24 03:04 -0700):
> I would have thought that ocamltest should always run tests in a clean
> environment (or something very close to that).

Well it creates its directories as it needs them so it can be run
anywhere I'd say.

xavierleroy commented:
@mshinwell:

> I would have thought that ocamltest should always run tests in a clean environment (or something very close to that).

That was my first reaction too, but for Windows/MSVC at least there are many environment variables that need to be set for ocamlopt to work.  So, maintaining the "clean environment" would be a major pain.

xavierleroy commented:
@shinderer:

> it creates its directories as it needs them so it can be run anywhere I'd say.

That's true, but we were talking about environment variables.

> Shall we add an "unset" directive to its DSL so that it can be guaranteed in a test that a given variable is not defined?

Sounds reasonable.  But @dra27's minimal fix is good enough, I think.

shindere commented:
Xavier Leroy (2019/10/24 05:30 -0700):
> @shinderer:
>
> > it creates its directories as it needs them so it can be run anywhere I'd say.
>
> That's true, but we were talking about environment variables.

Yeah, I realise the initial discussion was aobut environment variables,
but then reading @mshinwell's comment I was not sure what he was talking
aobut.

> > Shall we add an "unset" directive to its DSL so that it can be guaranteed in a test that a given variable is not defined?
>
> Sounds reasonable.  But @dra27's minimal fix is good enough, I think.

No: there is no test-specific Makefile anymore and even if there was
one, the test should work even if it's not run through a Makefile.

xavierleroy commented:
I was thinking of the top-level Makefile, testsuite/Makefile.

shindere commented:
Xavier Leroy (2019/10/24 07:42 -0700):
> I was thinking of the top-level Makefile, testsuite/Makefile.

That does not change anything to the story, I'd say. Youo always want to
be able to do something like

```bash
ocamltest foo/bar/baz.ml
```

When run that way, the test should, IMO, behave exactly the same way
than when run through a Makefile.

shindere commented:
So, what do we do on this, finally?

damiendoligez commented:
We should add an `unsetenv` directive to the DSL.

shindere commented:
Damien Doligez (2020/07/17 09:19 -0700):
> We should add an `unsetenv` directive to the DSL.

Will do.

dra27 committed:
Eliminate open Unix from testsuite

Mildly useful to be able to grep functions with Unix.foo

dra27 committed:
Harden win-unicode test against putenv failure

Rather than cleaning the environment in advance, use Unix.putenv twice,
so the environment could only possilby be initialised to one false
positive!

dra27 committed:
Add unset directive to ocamltest

unset erases the given variable from the environment.

dra27 committed:
Use unset in the testsuite to harden tests

dra27 commented:
Rebased and updated with the required changes. I've added a directive `unset` to ocamltest (I went with `unset` since we already have `set`). Probably because `Unix.environment` isn't available, ocamltest tracks environment changes internally with respect to an empty environment and only merges the two in the C stub. Unsetting environment variables therefore involves augmenting the type of `Environments.t` to be `string option Map.Make(Variables)`. Variables to be unset are signalled by `envp` entries which don't have an `=` sign. The Unix change is easy - we just use `unsetenv` after `fork` for these. The Windows stub is a bit tedious because it must construct the entire environment block in C - the loops used for that are not terribly efficient (it does a complete scan of the environment changes for each key in the environment) but the number of environment variables actually set in ocamltest is typically small (2-5ish), so I don't think it matters too much. The alternative would be to expose the `Unix.environment`y bit as a stub and do the computation completely in the OCaml side (so `Run_command.run` would receive the complete environment to use, rather than just the changes to it).

There is one open question for @shindere relating to a comment in https://github.com/ocaml/ocaml/pull/1599/files#diff-d619a1efbbbf9eda4513fd149ffb467cR192 - do you know offhand of any hooks which actually use the facility for deleting an environment variable (I can have a search if you're not sure) - I'm not sure looking at it that that it would have ever worked as intended, but it wasn't entirely clear to me.

I tested this for the original case here by breaking Unix.putenv (I made it a noop). If you then `export FOO2=BAR2` prior to `make -C testsuite one DIR=tests/lib-unix/win-env` then before this PR it would pass. With this PR, and the `unset FOO2` directive in the test, the test fails as you'd expect. There were several other tests which relied on variables not being defined in the environment externally (which used silly names to try to ensure this) and which now use less silly names and `unset` to guarantee it.

dra27 commented on ocamltest/run_win32.c:
```diff
@@ -175,18 +175,40 @@ static LPVOID prepare_environment(WCHAR **localenv)
+    /* Temporarily change the = to \0 for wcscmp */
+    *pos_eq = L'\0';
+    q = localenv;
+    while (*q != NULL) {
+      wchar_t *pos_eq2 = wcschr(*q, L'=');
+      /* Compare this name in localenv with the current one in processenv */
+      if (pos_eq2) *pos_eq2 = L'\0';
+      if (!wcscmp(*q, p)) copy = 0;
+      if (pos_eq2) *pos_eq2 = L'=';
+      q++;
+    }
```
This is clearly inefficient, although q should be short

dra27 commented on ocamltest/variables.ml:
```diff
@@ -33,7 +33,7 @@ exception Variable_already_registered of string
 exception No_such_variable of string
 
-let default_exporter varname value = Printf.sprintf "%s=%s" varname value
+let default_exporter varname value = (varname, value)
```
This is a belt and braces change - the code now relies on the environment representation `FOO=bar` so I didn't like that fact being exposed in custom exporters. This change means that a custom exporter can change both the name of the variable and the value (which is what it's actually used for) and then the formatting of it as `FOO=bar` is always done in `string_of_binding`.

damiendoligez replied:
Strictly speaking, this isn't quite right. As far as the Unix kernel is concerned, the environment is simply a list of strings. The `FOO=bar` representation is only a user-level convention. If a program makes some other use of the environment, you might not be able to test it correctly. Of course, this is all rather theoretical.

xavierleroy replied:
Quoting the Single Unix Specification:

> For a C-language program, an array of strings called the environment is made available when a process begins. [...]  These strings have the form name=value; names do not contain the character =. 

So, `VAR=value` may be "a user-level convention", but it's one that is set in stone in the standards.

xavierleroy commented:
I started to review but wasn't expecting so much code and got lost in the middle.

shindere commented:
This looks nice to me, thanks a lot.

Two remarks:

It feels to me that the functions defined by pattern-matching like expand
in environments.ml do not follow the style of the rest of the code because
the vertical bars are not indented. Could that be changed?

Also, perhaps rather than introducing this "XXX" comment this could be
clarified so that no such comment needs to be added?

dra27 committed:
Correct indentation of function clauses

dra27 committed:
Changes

dra27 commented:
> This looks nice to me, thanks a lot.

Thanks!

> Two remarks:
> 
> It feels to me that the functions defined by pattern-matching like expand
> in environments.ml do not follow the style of the rest of the code because
> the vertical bars are not indented. Could that be changed?

Indeed, I pushed an update.

> Also, perhaps rather than introducing this "XXX" comment this could be
> clarified so that no such comment needs to be added?

That's an open question - do you know offhand of any hooks which actually use the facility for deleting an environment variable (I can have a search if you're not sure) - I'm not sure looking at it that that it would have ever worked as intended, but it wasn't entirely clear to me. I certainly don't intend the comment in any form to be merged!

shindere commented:
David Allsopp (2020/09/09 01:50 -0700):
> > This looks nice to me, thanks a lot.
>
> Thanks!
>
> > Two remarks:
> >
> > It feels to me that the functions defined by pattern-matching like expand
> > in environments.ml do not follow the style of the rest of the code because
> > the vertical bars are not indented. Could that be changed?
>
> Indeed, I pushed an update.

Thanks. I thinkthere were several functions with that style so you may
want to double check.*

No, I can't think of hooks that do use the feature of removing variables
from the environment. I think this was implemented just to make sure
hooks have the  same abilities to modify the environment as any action
written in OCaml, but I don't thinkthis possibility is used at the
moment.

damiendoligez commented on ocamltest/environments.ml:
```diff
@@ -39,16 +41,48 @@ let rec expand env value =
+let append_to_system_env environment env =
+  (* Augment env with any bindings which are only in environment. This must be
+     done here as the Windows C implementation doesn't process multiple values
+     coming in in settings.envp. *)
```
```suggestion
     coming in settings.envp. *)
```

dra27 replied:
It was what I meant (read the second "in" as "via"), but it was very clunky - "~coming in~ in settings.envp." has the same meaning.

damiendoligez commented on ocamltest/environments.ml:
```diff
@@ -69,26 +103,29 @@ let safe_lookup variable env = match lookup variable env with
+let dump_assignment log = function
+  | (variable, Some value) ->
+    Printf.fprintf log "%s = %s\n%!" (Variables.name_of_variable variable) value
+  | (_, None) -> ()
```
`None` means "remove this variable from the environment", right? Shouldn't you print a line in this case? This will make the users' lives easier.

dra27 replied:
Yes - done!

(just for info, this is displayed when you use the `dumpenv` debugging action)

damiendoligez commented on ocamltest/environments.ml:
```diff
@@ -152,6 +189,7 @@ let rec apply_modifier environment = function
   | Add (variable, value) -> add variable value environment
   | Append (variable, value) -> append variable value environment
-  | Remove variable -> remove variable environment
+    (* XXX Is this used any where - does it actually work? *)
+  | Remove variable -> VariableMap.remove variable environment
```
Shouldn't this be updated to remove the variable from a pre-existing environment?
This is also a question for @shindere.

dra27 replied:
I went through this in more detail - this facility is used [`recvfrom_linux.ml`](https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/lib-unix/unix-socket/recvfrom_linux.ml) here:

https://github.com/ocaml/ocaml/blob/ddd9ec2a91ebfbd5bb18e75d2d37efac5e3d3bfd/testsuite/tests/lib-unix/unix-socket/is-linux.sh#L12-L15

It is important that this actually _removes_ the binding completely (as opposed to _adding_ an instruction to `unsetenv` a variable) or this code fails:

https://github.com/ocaml/ocaml/blob/ddd9ec2a91ebfbd5bb18e75d2d37efac5e3d3bfd/ocamltest/actions_helpers.ml#L221-L230

My conclusion is that what I've done is correct, but changing `Environments.remove` was the wrong way to expose it - in 39d29be I have left `Environments.remove` alone and instead added `Environments.unsetenv`.

So the `Remove` action affects the _ocamltest_ environment (i.e. it removes a binding from the alterations ocamltest expects to make to the system environment when running a command). The new `unset` instruction _adds_ an instruction to the _ocamltest_ environment whose effect is to call `unsetenv` on any variable in the _system_ environment.

damiendoligez commented on ocamltest/run_win32.c:
```diff
@@ -175,18 +175,40 @@ static LPVOID prepare_environment(WCHAR **localenv)
-    memcpy(r, p, l * sizeof(WCHAR));
+    /* Temporarily change the = to \0 for wcscmp */
+    *pos_eq = L'\0';
+    q = localenv;
+    while (*q != NULL) {
```
I'd use a `for` loop here.

dra27 replied:
This was in keeping with similar loops already present in the function, but I agree that they look clearer as `for` loops - I changed them all in de11da3.

damiendoligez commented on ocamltest/run_win32.c:
```diff
@@ -175,18 +175,40 @@ static LPVOID prepare_environment(WCHAR **localenv)
+    }
+    if (copy) {
+      /* This name is not marked for deletion/update in localenv, so copy */
+      *pos_eq = L'=';
```
I'd move this line up outside the `if`. The assigment to `*pos_eq` is supposed to be temporary, so it should be undone in all cases.

dra27 replied:
A good point, and changed.

damiendoligez commented on testsuite/tests/basic/opt_variants.ml:
```diff
@@ -1,7 +1,9 @@
-(* TEST *)
+(* TEST
+unset DOES_NOT_EXIST
```
```suggestion
unset FOO
```

dra27 replied:
I don't agree - the other instances set a variable where the purpose of using `unset` is to ensure that it was not set by the environment already. In 80a403, I changed the one other instance of an environment which is expected not to exist to be `DOES_NOT_EXIST` too. The other `FOO`, `FOO2`, etc. ones all get passed to `putenv` or similar, so the idea there is to ensure that the value they get really came from `putenv` and not randomly from the environment.

damiendoligez replied:
Makes sense.

damiendoligez commented on testsuite/tests/basic/opt_variants.ml:
```diff
@@ -1,7 +1,9 @@
 
 let () =
-  assert(Sys.getenv_opt "FOOBAR_UNLIKELY_TO_EXIST_42" = None);
+  assert(Sys.getenv_opt "DOES_NOT_EXIST" = None);
```
```suggestion
  assert(Sys.getenv_opt "FOO" = None);
```
For consistency with other tests.

dra27 replied:
(see above)

damiendoligez commented on testsuite/tests/lib-unix/common/test_unix_cmdline.ml:
```diff
@@ -31,27 +31,25 @@ all_modules= "test_unix_cmdline.ml"
 let run args =
-  let out, inp = pipe () in
-  let in_chan = in_channel_of_descr out in
+  let out, inp = Unix.pipe () in
```
Out of curiosity, is there a strong reason for these changes? (removal of `open Unix` and adding `Unix.` qualifiers all over the place)

dra27 replied:
grep'ing, as in the commit message - there were many more instances of qualified uses of `Unix`, that commit just made the tests consistently not use `open Unix` (or aliasing, but that wasn't used anyway) which meant I could see every use of `Unix.environment` or `Unix.getenv` with a simple fgrep...

damiendoligez commented on testsuite/tests/win-unicode/mltest.ml:
```diff
@@ -144,9 +144,18 @@ let test_open_in () =
 
 let test_getenv () =
+  let equiv l r =
+    assert (l = r);
+    l, r
+  in
   let doit key s =
     Unix.putenv key s;
-    Sys.getenv key, getenvironmentenv key
+    let l = equiv (Sys.getenv key) (getenvironmentenv key) in
+    let r =
+      Unix.putenv key (s ^ s);
+      equiv (Sys.getenv key) (getenvironmentenv key)
+    in
+      l, r
```
I don't understand the reason for this change.

dra27 replied:
This test sets a series of code-tortuously-named variables, explicitly testing that the Unicode version of `putenv` is actually working. If those environment variables have been pre-set with the expected values but `putenv` was broken, then the test would pass falsely.

Rather than having to be sure that UTF-8 names were working in `unset` in the `TEST` block, I altered the test itself so that it calls `putenv` twice, so regardless of the initial environment, a failing `putenv` should be detected, since the initial environment cannot yield two different values for the same variable (even on Windows!)

damiendoligez reviewed (approved):
This looks OK modulo a few questions and suggestions.

shindere commented:
Damien Doligez (2021/01/20 09:35 -0800):
> > +    (* XXX Is this used any where - does it actually work? *)
> +  | Remove variable -> VariableMap.remove variable environment
>
> Shouldn't this be updated to remove the variable from a pre-existing environment?
> This is also a question for @shindere.

Do you mean from the system environemnt? Then yes, probably so.

gasche commented:
I came across this PR in my sweep of old issues. It looks fine: there was a review not so long ago (January 2021), and we are well within an OCaml-reasonable delay for @dra27 to respond and integrate changes.

dra27 committed:
Environments.remove -> Environments.unsetenv

dra27 committed:
Use for instead of while for stepwise loops

dra27 committed:
Always restore alteration to environment string

dra27 committed:
Use DOES_NOT_EXIST to clarify expectation

dra27 commented:
Thank you for the review, @damiendoligez and sorry it's taken quite so long to address the points. I've rebased this, but the changes are all in new commits (the most recent 4) to aid re-review.

The `XXX` comment is now properly removed - assuming remains approved, I'd like to tidy up the commit history prior to merging, but it's otherwise ready from my perspective, assuming CI doesn't now do throw up something strange.

damiendoligez commented:
I went through the latest changes and my approval still stands.

gasche commented:
Thanks! Merged.

dra27 commented:
I had been going to tidy the commits this morning, but thanks for the merge, regardless 🙂 Thank you for the time reviewing it, @damiendoligez and @shindere!
